rust/main/agents/relayer/src/msg/pending_message.rs

```diff
@@ -649,6 +643,71 @@ impl PendingMessage {
         PendingOperationStatus::FirstPrepareAttempt
     }
 
+    /// A preflight check to see if a message could possibly meet
+    /// a gas payment requirement prior to undertaking expensive operations
+    /// like metadata building or gas estimation.
+    /// If the message does not meet the gas payment requirement,
+    /// Err(PendingOperationResult) is returned, with the PendingOperationResult intended
+    /// to be propagated up by the prepare fn.
+    async fn meets_gas_payment_requirement_preflight_check(
+        &mut self,
+    ) -> Result<(), PendingOperationResult> {
+        // We test if the message may meet the gas payment requirement
+        // with the most simple tx cost estimate: one that has zero cost
+        // whatsoever. If the message does not meet the gas payment requirement
+        // with zero cost, we can skip the metadata building and gas estimation
+        // altogether. This covers the case of a message that did not pay our IGP,
+        // which may violate the gas payment enforcement policies depending on
+        // the configuration, but also allows us to be tolerant of the configuration
+        // allowing no payment at all.
+        let zero_cost = TxCostEstimate {
+            gas_limit: U256::zero(),
+            gas_price: FixedPointNumber::zero(),
+            l2_gas_limit: None,
+        };
+
+        self.meets_gas_payment_requirement(&zero_cost)
+            .await
+            // No need to surface the gas_limit if it's met, an Ok(()) suffices
+            .map(|_| ())
+    }
+
+    /// Returns the gas limit if the message meets the gas payment requirement,
+    /// otherwise returns an Err(PendingOperationResult), with the PendingOperationResult intended
+    /// to be propagated up by the prepare fn.
+    async fn meets_gas_payment_requirement(
+        &mut self,
+        tx_cost_estimate: &TxCostEstimate,
+    ) -> Result<U256, PendingOperationResult> {
```

Comment:
IMO this overloads the meaning of `Result`, since the `Err` variant isn't really an error, which can be confusing. Would be cleaner to define an enum
```rust
enum IgpPolicyOutcome {
    MeetsPolicy(U256),
    PolicyNotMet(PendingOperationResult)
}
```
```suggestion
    ) -> IgpPolicyOutcome {
```

Reply:
yeah you're def right, will do that